core/src/main/java/jenkins/telemetry/impl/StaplerDispatches.java

```diff
@@ -0,0 +1,119 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.telemetry.impl;
+
+import hudson.Extension;
+import hudson.PluginWrapper;
+import hudson.model.UsageStatistics;
+import hudson.util.VersionNumber;
+import jenkins.model.Jenkins;
+import jenkins.telemetry.Telemetry;
+import jenkins.util.SystemProperties;
+import net.sf.json.JSONObject;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.EvaluationTrace;
+import org.kohsuke.stapler.StaplerRequest;
+
+import javax.annotation.Nonnull;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.time.LocalDate;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TimeZone;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+/**
+ * Telemetry implementation gathering information about Stapler dispatch routes.
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public class StaplerDispatches extends Telemetry {
+    @Nonnull
+    @Override
+    public LocalDate getStart() {
+        return LocalDate.of(2018, 10, 10);
+    }
+
+    @Nonnull
+    @Override
+    public LocalDate getEnd() {
+        return LocalDate.of(2019, 2, 1);
+    }
+
+    @Nonnull
+    @Override
+    public String getDisplayName() {
+        return "Stapler request handling";
+    }
+
+    @Nonnull
+    @Override
+    public JSONObject createContent() {
+        Map<String, Object> info = new TreeMap<>();
+        info.put("components", buildComponentInformation());
+        info.put("dispatches", buildDispatches());
+
+        return JSONObject.fromObject(info);
+    }
+
+    private Object buildDispatches() {
+        Set<String> currentTraces = traces;
+        traces = new TreeSet<>();
+        return currentTraces;
+    }
+
+    private Object buildComponentInformation() {
+        Map<String, String> components = new TreeMap<>();
+        VersionNumber core = Jenkins.getVersion();
+        components.put("jenkins-core", core == null ? "" : core.toString());
+
+        for (PluginWrapper plugin : Jenkins.get().pluginManager.getPlugins()) {
+            if (plugin.isActive()) {
+                components.put(plugin.getShortName(), plugin.getVersion());
+            }
+        }
+        return components;
+    }
+
+    @MetaInfServices
+    public static class StaplerTrace extends EvaluationTrace.ApplicationTracer {
+
+        @Override
+        protected void record(StaplerRequest staplerRequest, String s) {
+            if (UsageStatistics.DISABLED || !Jenkins.get().isUsageStatisticsCollected()) {
```

Comment:
I'm assuming this is just an optimization, since Telemetry is not reported regardless when these settings are present, correct?

Reply:
Yes. This covers two potential issues:

* Accumulation of trace information without ever being cleared out on submission, as there is no submission.
* Having disabled usage statistics, then enabling it. User expectation is likely that the next submission will only contain data collected from that point forward, not any older.

Reply:
I do something similar in https://github.com/jenkinsci/jenkins/pull/3687/files#diff-f2ee06c0611e993efba3c8330e1b98feR114 but that PR adds new core API to do the check so I didn't do it here to prevent merge conflicts. Should probably be cleaned up later.

Reply:
Please add a `TODO` comment here reminding people to use that API once merged.

Reply:
Done.